datagateway_api/src/search_api/models.py

```diff
@@ -1,40 +1,147 @@
-from abc import ABC, abstractclassmethod
-from datetime import datetime
+import abc
+from abc import ABC
+from datetime import datetime, timezone
+import sys
 from typing import ClassVar, List, Optional, Union
 
-from pydantic import (
-    BaseModel,
-    Field,
-    root_validator,
-    StrictBool,
-    StrictFloat,
-    StrictInt,
-    StrictStr,
-)
+from dateutil.relativedelta import relativedelta
+from pydantic import BaseModel, Field, ValidationError, validator
+from pydantic.error_wrappers import ErrorWrapper
+
+from datagateway_api.src.search_api.panosc_mappings import mappings
+
+
+def _get_icat_field_value(icat_field_name, icat_data):
+    icat_field_name = icat_field_name.split(".")
+    for field_name in icat_field_name:
+        if isinstance(icat_data, list):
+            values = []
+            for data in icat_data:
+                value = _get_icat_field_value(field_name, data)
+                if isinstance(value, list):
+                    values.extend(value)
+                else:
+                    values.append(value)
+
+            icat_data = values
+
+        if isinstance(icat_data, dict):
+            icat_data = icat_data[field_name]
+
+    return icat_data
 
 
 class PaNOSCAttribute(ABC, BaseModel):
-    @abstractclassmethod
-    def from_icat(self):
-        pass
+    @classmethod
+    @abc.abstractmethod
+    def from_icat(cls, icat_data, required_related_fields):  # noqa: B902, N805
+        model_fields = cls.__fields__
+
+        model_data = {}
+        for field in model_fields:
+            # Some fields have aliases so we must use them when creating a model instance.
+            # If a field does not have an alias then the `alias` property holds the name
+            # of the field
+            field_alias = cls.__fields__[field].alias
+
+            panosc_entity_name, icat_field_name = mappings.get_icat_mapping(
+                cls.__name__, field_alias,
+            )
+
+            if not isinstance(icat_field_name, list):
+                icat_field_name = [icat_field_name]
+
+            field_value = None
+            for field_name in icat_field_name:
+                try:
+                    value = _get_icat_field_value(field_name, icat_data)
+                    if value:
+                        field_value = value
+                        break
```

Comment:
Why do we need `value`, can't we just put the result of `_get_icat_field_value()` into `field_value`? Then you could remove line 59 and just have the break in the conditional. I've just thrown together a suggestion but you should test this as there may be a reason why this won't actually work.

```suggestion
                    field_value = _get_icat_field_value(field_name, icat_data)
                    if field_value:
                        break
```